src/main/java/com/zaxxer/nuprocess/internal/BasePosixProcess.java

```diff
@@ -82,6 +92,26 @@
 
    private Pointer posix_spawn_file_actions;
 
+   // Launches threads under which changes to cwd do not affect the cwd of the process.
+   public static class LinuxCwdChangeableThreadFactory implements ThreadFactory {
+      private final AtomicInteger threadCount = new AtomicInteger(0);
+
+      @Override
+      public Thread newThread(final Runnable r) {
+         Runnable unshareCwdThenSpawn = new Runnable() {
+            @Override
+            public void run() {
+              int rc = LinuxLibC.unshare(LinuxLibC.CLONE_FS);
```

Comment:
Will clean up, thanks!

unshare(CLONE_FS) makes a copy of the existing settings including the
chroot, so it doesn't introduce a security flaw. (It would be a massive
security flaw if it changed the chroot!)

I'll document who has to catch the exceptions.

2015年9月5日(土) 0:10 Brett Wooldridge <notifications@github.com>:

> In src/main/java/com/zaxxer/nuprocess/internal/BasePosixProcess.java
> <https://github.com/brettwooldridge/NuProcess/pull/32#discussion_r38808507>
> :
>
> > @@ -82,6 +92,26 @@
> >
> >     private Pointer posix_spawn_file_actions;
> >
> > +   // Launches threads under which changes to cwd do not affect the cwd of the process.
> > +   public static class LinuxCwdChangeableThreadFactory implements ThreadFactory {
> > +      private final AtomicInteger threadCount = new AtomicInteger(0);
> > +
> > +      @Override
> > +      public Thread newThread(final Runnable r) {
> > +         Runnable unshareCwdThenSpawn = new Runnable() {
> > +            @Override
> > +            public void run() {
> > +              int rc = LinuxLibC.unshare(LinuxLibC.CLONE_FS);
>
> One concern about unshare(). The doc says:
>
> CLONE_FS
> Unshare filesystem attributes, so that the calling process no longer
> shares its root directory (chroot(2)), current directory (chdir(2)), or
> umask (umask(2)) attributes with any other process.
>
> Does this allow the spawned process to escape a chroot jail
> <http://docs.oracle.com/cd/E37670_01/E36387/html/ol_cj_sec.html>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/brettwooldridge/NuProcess/pull/32/files#r38808507>.
>